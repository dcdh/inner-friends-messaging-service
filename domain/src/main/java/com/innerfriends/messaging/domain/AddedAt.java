package com.innerfriends.messaging.domain;

import java.time.ZonedDateTime;
import java.util.Objects;

public final class AddedAt {

    private final ZonedDateTime at;

    public AddedAt(final EventAt eventAt) {
        this(eventAt.at());
    }

    public AddedAt(final PostedAt postedAt) {
        this(postedAt.at());
    }

    public AddedAt(final ZonedDateTime at) {
        this.at = at;
    }

    public ZonedDateTime at() {
        return at;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AddedAt)) return false;
        AddedAt addedAt = (AddedAt) o;
        return Objects.equals(at, addedAt.at);
    }

    @Override
    public int hashCode() {
        return Objects.hash(at);
    }

    @Override
    public String toString() {
        return "AddedAt{" +
                "at=" + at +
                '}';
    }
}
