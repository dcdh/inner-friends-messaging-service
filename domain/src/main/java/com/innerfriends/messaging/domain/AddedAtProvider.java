package com.innerfriends.messaging.domain;

public interface AddedAtProvider {

    AddedAt generate();

}
