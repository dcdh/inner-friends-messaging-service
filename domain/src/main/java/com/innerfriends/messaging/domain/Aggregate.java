package com.innerfriends.messaging.domain;

import java.util.Objects;
import java.util.concurrent.Callable;
import java.util.function.Consumer;

public abstract class Aggregate {

    private Long version;

    public Aggregate(final Long version) {
        this.version = Objects.requireNonNull(version);
        if (version < 0) {
            throw new IllegalStateException();
        }
    }

    protected void apply(final Callable apply) {
        try {
            apply.call();
        } catch (final Exception exception) {
            throw new FailedToMutateAggregateException(exception);
        }
        version++;
    }

    public Long version() {
        return version;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof Aggregate)) return false;
        final Aggregate aggregate = (Aggregate) o;
        return Objects.equals(version, aggregate.version);
    }

    @Override
    public int hashCode() {
        return Objects.hash(version);
    }

    @Override
    public String toString() {
        return "Aggregate{" +
                "version=" + version +
                '}';
    }
}
