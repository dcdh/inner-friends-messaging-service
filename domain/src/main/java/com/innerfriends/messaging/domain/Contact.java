package com.innerfriends.messaging.domain;

import java.util.Objects;

public final class Contact {

    private final ContactIdentifier contactIdentifier;
    private final AddedAt addedAt;

    public Contact(final ContactIdentifier contactIdentifier, final AddedAt addedAt) {
        this.contactIdentifier = Objects.requireNonNull(contactIdentifier);
        this.addedAt = Objects.requireNonNull(addedAt);
    }

    public ContactIdentifier contactIdentifier() {
        return contactIdentifier;
    }

    public AddedAt addedAt() {
        return addedAt;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Contact)) return false;
        Contact contact = (Contact) o;
        return Objects.equals(contactIdentifier, contact.contactIdentifier) &&
                Objects.equals(addedAt, contact.addedAt);
    }

    @Override
    public int hashCode() {
        return Objects.hash(contactIdentifier, addedAt);
    }

    @Override
    public String toString() {
        return "Contact{" +
                "contactIdentifier=" + contactIdentifier +
                ", addedAt=" + addedAt +
                '}';
    }
}
