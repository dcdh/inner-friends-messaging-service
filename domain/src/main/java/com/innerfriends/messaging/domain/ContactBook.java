package com.innerfriends.messaging.domain;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public final class ContactBook extends Aggregate {

    private final Owner owner;
    private final List<Contact> contacts;

    public ContactBook(final Owner owner,
                       final List<Contact> contacts,
                       final Long version) {
        super(version);
        this.owner = Objects.requireNonNull(owner);
        this.contacts = Objects.requireNonNull(contacts);
    }

    public ContactBook(final Owner owner) {
        this(owner, new ArrayList<>(), 0l);
    }

    public ContactBook addNewContact(final ContactIdentifier contactIdentifier, final AddedAt addedAt) {
        this.apply(() -> this.contacts.add(new Contact(contactIdentifier, addedAt)));
        return this;
    }

    public List<ContactIdentifier> recentContacts(final Integer nbOfContactToReturn) {
        return contacts.stream()
                .sorted((e1, e2) -> e2.addedAt().at().compareTo(e1.addedAt().at()))
                .limit(nbOfContactToReturn)
                .map(Contact::contactIdentifier)
                .collect(Collectors.toUnmodifiableList());
    }

    public List<ContactIdentifier> allContacts() {
        return contacts.stream()
                .sorted(Comparator.comparing(e -> e.contactIdentifier().identifier().identifier()))
                .map(Contact::contactIdentifier)
                .collect(Collectors.toUnmodifiableList());
    }

    public boolean hasContact(final ParticipantIdentifier participantIdentifier) {
        return contacts.stream()
                .anyMatch(contact -> contact.contactIdentifier().isEqualTo(participantIdentifier));
    }

    public Owner owner() {
        return owner;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof ContactBook)) return false;
        if (!super.equals(o)) return false;
        final ContactBook that = (ContactBook) o;
        return Objects.equals(owner, that.owner) &&
                Objects.equals(contacts, that.contacts);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), owner, contacts);
    }

    @Override
    public String toString() {
        return "ContactBook{" +
                "owner=" + owner +
                ", contacts=" + contacts +
                "} " + super.toString();
    }
}
