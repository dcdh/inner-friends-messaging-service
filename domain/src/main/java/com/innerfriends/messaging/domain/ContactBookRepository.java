package com.innerfriends.messaging.domain;

public interface ContactBookRepository {

    ContactBook getByOwner(Owner owner) throws NoContactBookFoundException;

    void save(ContactBook contactBook) throws NoContactBookFoundException;

}
