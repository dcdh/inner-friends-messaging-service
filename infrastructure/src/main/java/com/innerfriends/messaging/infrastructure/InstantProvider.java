package com.innerfriends.messaging.infrastructure;

import javax.enterprise.context.ApplicationScoped;
import java.time.Instant;

@ApplicationScoped
public class InstantProvider {

    public Instant now() {
        return Instant.now();
    }

}
