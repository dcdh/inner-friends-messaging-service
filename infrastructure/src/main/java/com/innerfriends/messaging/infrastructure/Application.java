package com.innerfriends.messaging.infrastructure;

import com.innerfriends.messaging.domain.*;
import com.innerfriends.messaging.domain.usecase.*;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.inject.Produces;
import java.time.ZonedDateTime;
import java.util.UUID;

public class Application {

    @Produces
    @ApplicationScoped
    public AddedAtProvider addedAtProviderProducer() {
        return () -> new AddedAt(ZonedDateTime.now());
    }

    @Produces
    @ApplicationScoped
    public ConversationIdentifierProvider conversationIdentifierProvider() {
        return new ConversationIdentifierProvider() {
            @Override
            public ConversationIdentifier generate(final OpenedBy openedBy) {
                return new ConversationIdentifier(openedBy.identifier().identifier() + "_" + UUID.randomUUID());
            }
        };
    }

    @Produces
    @ApplicationScoped
    public PostedAtProvider postedAtProvider() {
        return new PostedAtProvider() {
            @Override
            public PostedAt now() {
                return new PostedAt(ZonedDateTime.now());
            }
        };
    }

    @Produces
    @ApplicationScoped
    public AddContactIntoContactBookUseCase addContactIntoContactBookUseCaseProducer(final ContactBookRepository contactBookRepository,
                                                                                     final AddedAtProvider addedAtProvider) {
        return new AddContactIntoContactBookUseCase(contactBookRepository, addedAtProvider);
    }

    @Produces
    @ApplicationScoped
    public CreateContactBookUseCase createContactBookUseCaseProducer(final ContactBookRepository contactBookRepository) {
        return new CreateContactBookUseCase(contactBookRepository);
    }

    @Produces
    @ApplicationScoped
    public ListAllContactsUseCase listAllContactsUseCaseProducer(final ContactBookRepository contactBookRepository) {
        return new ListAllContactsUseCase(contactBookRepository);
    }

    @Produces
    @ApplicationScoped
    public ListConversationsUseCase listConversationsUseCaseProducer(final ConversationRepository conversationRepository) {
        return new ListConversationsUseCase(conversationRepository);
    }

    @Produces
    @ApplicationScoped
    public ListMessagesInConversationUseCase listMessagesInConversationUseCaseProducer(final ConversationRepository conversationRepository) {
        return new ListMessagesInConversationUseCase(conversationRepository);
    }

    @Produces
    @ApplicationScoped
    public ListRecentContactsUseCase listRecentContactsUseCaseProducer(final ContactBookRepository contactBookRepository) {
        return new ListRecentContactsUseCase(contactBookRepository);
    }

    @Produces
    @ApplicationScoped
    public PostNewMessageToConversationUseCase postNewMessageToConversationUseCase(final ConversationRepository conversationRepository,
                                                                                   final PostedAtProvider postedAtProvider) {
        return new PostNewMessageToConversationUseCase(conversationRepository, postedAtProvider);
    }

    @Produces
    @ApplicationScoped
    public OpenANewConversationUseCase startConversationUseCaseProducer(final ConversationRepository conversationRepository,
                                                                        final ContactBookRepository contactBookRepository,
                                                                        final ConversationIdentifierProvider conversationIdentifierProvider,
                                                                        final PostedAtProvider postedAtProvider) {
        return new OpenANewConversationUseCase(conversationRepository, contactBookRepository, conversationIdentifierProvider, postedAtProvider);
    }

}
