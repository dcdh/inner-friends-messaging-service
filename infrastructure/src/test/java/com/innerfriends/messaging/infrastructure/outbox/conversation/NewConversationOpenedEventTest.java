package com.innerfriends.messaging.infrastructure.outbox.conversation;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.innerfriends.messaging.domain.*;
import com.innerfriends.messaging.infrastructure.InstantProvider;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;

@ExtendWith(MockitoExtension.class)
public class NewConversationOpenedEventTest {

    private InstantProvider instantProvider;

    private static final ObjectMapper objectMapper = new ObjectMapper();

    @BeforeEach
    public void setup() {
        this.instantProvider = mock(InstantProvider.class);
    }

    @Test
    public void should_return_expected_event() {
        // Given
        final ConversationIdentifier conversationIdentifier = new ConversationIdentifier("Mario-azerty");
        final Conversation conversation = new Conversation(conversationIdentifier,
                new Message(new From("Mario"), buildPostedAt(2), new Content("Hi Peach how are you ?")),
                List.of(new ParticipantIdentifier("Mario"), new ParticipantIdentifier("Peach")));
        doReturn(Instant.ofEpochSecond(1)).when(instantProvider).now();

        // When
        final NewConversationOpenedEvent newConversationOpenedEvent = NewConversationOpenedEvent.of(conversation, objectMapper, instantProvider);

        // Then
        assertThat(newConversationOpenedEvent.getAggregateId()).isEqualTo("Mario-azerty");
        assertThat(newConversationOpenedEvent.getAggregateType()).isEqualTo("Conversation");
        assertThat(newConversationOpenedEvent.getTimestamp()).isEqualTo(Instant.ofEpochSecond(1));
        assertThat(newConversationOpenedEvent.getType()).isEqualTo("NewConversationOpened");
        assertThat(newConversationOpenedEvent.getPayload().toString()).isEqualTo("{\"conversationIdentifier\":\"Mario-azerty\",\"version\":0,\"firstMessage\":{\"from\":\"Mario\",\"content\":\"Hi Peach how are you ?\",\"postedAt\":\"2021-10-02T00:00:00+02:00[Europe/Paris]\"},\"participantsIdentifier\":[\"Mario\",\"Peach\"]}");
    }

    private PostedAt buildPostedAt(final Integer day) {
        return new PostedAt(
                ZonedDateTime.of(2021, 10, day, 0, 0, 0, 0, ZoneId.of("Europe/Paris")));
    }

}
