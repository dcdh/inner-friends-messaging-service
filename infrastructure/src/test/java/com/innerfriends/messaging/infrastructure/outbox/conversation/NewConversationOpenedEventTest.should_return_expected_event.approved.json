{
  "conversationIdentifier": "Mario-azerty",
  "version": 0,
  "firstMessage": {
    "from": "Mario",
    "content": "Hi Peach how are you ?",
    "postedAt": "2021-10-02T00:00:00+02:00[Europe/Paris]"
  },
  "participantsIdentifier": [
    "Mario",
    "Peach"
  ]
}